qcodes/instrument_drivers/Keysight/keysight_34980a_submodules.py

```diff
@@ -0,0 +1,384 @@
+import logging
+import warnings
+from functools import wraps
+import re
+import numpy as np
+from qcodes import VisaInstrument, InstrumentChannel, validators
+from typing import Union, List, Tuple, Optional, Callable
+
+LOGGER = logging.getLogger()
+
+
+def post_execution_status_poll(func: Callable) -> Callable:
+    """
+    Generates a decorator that clears the instrument's status registers
+    before executing the actual call and reads the status register after the
+    function call to determine whether an error occurs.
+
+    Args:
+        func: function to wrap
+    """
+    @wraps(func)
+    def wrapper(self, *args, **kwargs):
+        self.clear_status()
+        retval = func(self, *args, **kwargs)
+
+        stb = self.get_status()
+        if stb:
+            warnings.warn(f"Instrument status byte indicates an error occurred "
+                          f"(value of STB was: {stb})! Use `get_error` method "
+                          f"to poll error message.",
+                          stacklevel=2)
+        return retval
+
+    return wrapper
+
+
+class KeysightSubModule(InstrumentChannel):
```

Comment:
It's a great idea to have the SubMod class for general modules and a subclass SwitchMatrixSubMod for switch matrix modules. (This will solve more than one problem!)
However, just to confirm, the abstract `validate_value` and `to_channel_list` will still be in the SwitchMatrixSubMod, is that ok?

Reply:
> the abstract validate_value and to_channel_list will still be in the SwitchMatrixSubMod, is that ok?

sure! ( if those turn out to be really needed ;) )